### PR TITLE
Workflow backends: ignore on hot reloading

### PR DESCRIPTION
**Pull request by JoshVanL**
Updates the hot reloading reconciler so that Daprd will log and error and ignore workflow backends when they are hot reloaded.

Adds integration tests for daprd ensuring workflow backends can be loaded on boot.

Adds tests to ensure daprd will log an error on a workflow backend being hot reloaded.

Adds tests to ensure multiple workflow backends cannot be loaded at once.

closes https://github.com/dapr/dapr/issues/7409

### CODE REVIEW

File: pkg/runtime/hotreload/reconciler/reconciler.go [hunk 1]
```diff
@@ -140,34 +140,42 @@ func (r *Reconciler[T]) reconcile(ctx context.Context, result *differ.Result[T])
 		{result.Updated, operatorpb.ResourceEventType_UPDATED},
 		{result.Created, operatorpb.ResourceEventType_CREATED},
 	} {
-		wg.Add(len(group.resources))
+		errCh := make(chan error)
```
Comment (ItalyPaleAle) [inline pkg/runtime/hotreload/reconciler/reconciler.go:143]:
```suggestion
		errCh := make(chan error, len(group.resources))
```

File: pkg/runtime/hotreload/reconciler/reconciler.go [hunk 2]
```diff
@@ -140,34 +140,42 @@ func (r *Reconciler[T]) reconcile(ctx context.Context, result *differ.Result[T])
...
+				errCh <- r.handleEvent(ctx, &loader.Event[T]{
 					Type:     eventType,
 					Resource: resource,
 				})
 			}(resource, group.eventType)
 		}
-		wg.Wait()
+
+		errs := make([]error, 0, len(group.resources))
+		for range group.resources {
```
Comment (ItalyPaleAle) [inline pkg/runtime/hotreload/reconciler/reconciler.go:154]:
```suggestion
		for i := 0; i < len(group.resources); i++ {
```

File: pkg/runtime/compstore/compstore.go [hunk 3]
```diff
@@ -48,7 +49,7 @@ type ComponentStore struct {
 	pubSubs                 map[string]PubsubItem
 	topicRoutes             map[string]TopicRoutes
 	workflowComponents      map[string]workflows.Workflow
-	workflowBackends        map[string]wfbe.WorkflowBackend
+	workflowBackends        map[string]backend.Backend
```
Comment (ItalyPaleAle) [inline pkg/runtime/compstore/compstore.go:52]: Why this change?

File: pkg/runtime/hotreload/reconciler/component.go [hunk 4]
```diff
@@ -36,10 +37,14 @@ type component struct {
 // the generic reconciler.
 //
 //nolint:unused
-func (c *component) update(ctx context.Context, comp componentsapi.Component) {
+func (c *component) update(ctx context.Context, comp componentsapi.Component) error {
+	if strings.HasPrefix(comp.Spec.Type, "workflowbackend.") {
+		return fmt.Errorf("attempting to HotReload a workflowbackend component which is not supported: %s", comp.LogName())
```
Comment (ItalyPaleAle) [inline pkg/runtime/hotreload/reconciler/component.go:42]:
```suggestion
		return fmt.Errorf("attempting to hot-reloading a workflowbackend component which is not supported: %s", comp.LogName())
```

File: pkg/runtime/hotreload/reconciler/component.go [hunk 5]
```diff
@@ -48,32 +53,40 @@ func (c *component) update(ctx context.Context, comp componentsapi.Component) {
...
 
 //nolint:unused
-func (c *component) delete(comp componentsapi.Component) {
+func (c *component) delete(comp componentsapi.Component) error {
+	if strings.HasPrefix(comp.Spec.Type, "workflowbackend.") {
+		return fmt.Errorf("attempting to HotReload a workflowbackend component which is not supported: %s", comp.LogName())
+	}
+
 	if err := c.proc.Close(comp); err != nil {
 		log.Errorf("error closing deleted component: %w", err)
```
Comment (ItalyPaleAle) [inline pkg/runtime/hotreload/reconciler/component.go:88]: Logs don't support `%w`
```suggestion
		log.Errorf("error closing deleted component: %v", err)
```

File: pkg/runtime/hotreload/reconciler/component.go [hunk 6]
```diff
@@ -48,32 +53,40 @@ func (c *component) update(ctx context.Context, comp componentsapi.Component) {
...
 		if differ.AreSame(oldComp, comp) {
 			log.Debugf("Component update skipped: no changes detected: %s", comp.LogName())
-			return
+			return nil
 		}
 
 		log.Infof("Closing existing Component to reload: %s", oldComp.LogName())
 		// TODO: change close to accept pointer
 		if err := c.proc.Close(oldComp); err != nil {
 			log.Errorf("error closing old component: %w", err)
```
Comment (ItalyPaleAle) [inline pkg/runtime/hotreload/reconciler/component.go:62]:
```suggestion
			log.Errorf("error closing old component: %v", err)
```

File: pkg/runtime/compstore/compstore.go (hunk 3 above)
Comment (JoshVanL) [reply pkg/runtime/compstore/compstore.go:52]: This is the correct type for the workflow backend implementation interface. Notice the other component types are using the interface types defined in `components-contrib` however backends is using `durabletask-go` directly. This wasn't picked up before as this field wasn't previously being used (`workflowbackend.go` didn't exist). `wfbe.WorkflowBackend` is a factory func type.

File: pkg/runtime/hotreload/reconciler/reconciler.go (hunk 2 above)
Comment (JoshVanL) [reply pkg/runtime/hotreload/reconciler/reconciler.go:154]: What is the benefit of creating the variable `i` here?

File: pkg/runtime/processor/wfbackend/wfbackend.go [hunk 7]
```diff
@@ -56,18 +57,18 @@ func New(opts Options) *workflowBackend {
 	}
 }
 
-func (wfbe *workflowBackend) Init(ctx context.Context, comp compapi.Component) error {
-	wfbe.lock.Lock()
-	defer wfbe.lock.Unlock()
+func (w *workflowBackend) Init(ctx context.Context, comp compapi.Component) error {
```
[the remark this answers is not in the document]
Comment (JoshVanL) [reply pkg/runtime/processor/wfbackend/wfbackend.go:60]: By convention we use single letter variable names for method receiver variables. This gives overhead to the reader as it is unexpected. `wfbe` initially tripped my up as I was confused as to where this variable was coming from until I realized it was the receiver.

**Review by artursouza — APPROVED** [review]

**Review by artursouza — APPROVED** [review]
